src/redisearch_rs/rlookup/src/lookup.rs

```diff
@@ -997,6 +994,107 @@ impl<'a> RLookup<'a> {
         // Safety: We treat the pointer as pinned internally and safe Rust cannot move out of the returned immutable reference.
         Some(unsafe { Pin::into_inner_unchecked(key.into_ref()) })
     }
+
+    // ===== Load key from redis keyspace (include known information on the key, fail if already loaded) =====
+
+    pub fn get_key_load(
+        &mut self,
+        name: &'a CStr,
+        field_name: &'a CStr,
+        mut flags: RLookupKeyFlags,
+    ) -> Option<&RLookupKey<'a>> {
+        // remove all flags that are not relevant to getting a key
+        flags &= GET_KEY_FLAGS;
+
+        // 1. if the key is already loaded, or it has created by earlier RP for writing, return NULL (unless override was requested)
+        // 2. create a new key with the name of the field, and mark it as doc-source.
+        // 3. if the key is in the schema, mark it as schema-source and apply all the relevant flags according to the field spec.
+        // 4. if the key is "loaded" at this point (in schema, sortable and un-normalized), create the key but return NULL
+        //    (no need to load it from the document).
+
+        // Safety: The non-lexical lifetime analysis of current Rust, incorrectly handles borrows in early
+        // return statements, expanding the borrow of `self` to the scope of the entire method. This is
+        // obviously not correct as `key` is either found and borrowed in which case we never reach the code below
+        // this early return OR it is *not* found and *not* borrowed which means the code below is fine to
+        // borrow self again. The current compiler is not smart enough to get this though, so we create a disjoint
+        // borrow below.
+        // TODO remove once <https://github.com/rust-lang/rust/issues/54663> is fixed.
+        let me = unsafe { &mut *(self as *mut Self) };
```

Comment:
This is dynamite waiting to explode. Can we restructure the rest of the method to work around the issue and avoid this unsafe block?

Reply:
unsure, fundamentally these methods are all `get_or_insert` shaped. To my knowledge no easy solution exists for this that's not polonius. That only leaves large restructuring as an option, which idk... also doesn't feel great.